File: Data_Structure/LinkedList/Reverse_Singly_LinkedList_2.c
//Reverse a linked list without actually reversing

#include<stdio.h>
#include<stdlib.h>

struct node
{
    int data;
    struct node * next;
};


void linkedListTraversal(struct node* ptr)
{
    while(ptr != NULL)
    {
        printf("%d ", ptr -> data);
        ptr = ptr -> next;
    }
}

void reverseLinkedList(struct node* ptr)
{
    if(ptr == NULL)
    {
        return;
    }   
    reverseLinkedList(ptr -> next);
    printf("%d ", ptr->data);
}


int main()
{
    struct node *head;
    struct node *second;
    struct node *third;
    struct node *fourth;
    
    // Allocate memory for nodes in the linked list in Heap
    head = (struct node*)malloc(sizeof(struct node));
    second = (struct node*)malloc(sizeof(struct node));
    third = (struct node*)malloc(sizeof(struct node));
    fourth = (struct node*)malloc(sizeof(struct node));
    
    // Link first and second nodes
    head -> data = 37;
    head -> next = second;
    
    second -> data = 62;
    second -> next = third;
    
    third -> data = 26;
    third -> next = fourth;
    
    fourth -> data = 93;
    fourth -> next = NULL;
    
    printf("List before reversing\n");
    linkedListTraversal(head);
    
    printf("\n\nList after reversing\n");
    reverseLinkedList(head);
    return 0;
}

File: Data_Structure/DS_Using_Python/Reverse_Singly_LinkedList_2.py
#Reverse a linked list without actually reversing

class Node:
	def __init__(self, data):
		self.data = data
		self.next = None
		
class SinglyLinkedList:
	def __init__(self):
		self.head = None
		
	def printList(self):
		if self.head is None:
			print('List has no element')
			return
		else:
			temp = self.head
			while(temp):
				print(temp.data, end=' ')
				temp = temp.next
				
	def size(self):
		if self.head == None:
			return 0 
		else:
			temp = self.head
			c = 0
			while(temp is not None):
				c += 1
				temp = temp.next
			return c
		
	def insert(self, data):
		newNode = Node(data)
		newNode.next = self.head
		self.head = newNode
		
	def reverse(self, temp):
			if temp == None:
				return
			self.reverse(temp.next)
			print(temp.data,end=' ')
				
llist = SinglyLinkedList()
llist.insert(35)
llist.insert(2)
llist.insert(97)
llist.insert(53)

print('\nLinked list before reversing')
llist.printList()

print('\n\nLinked list after reversing')
llist.reverse(llist.head)

if llist.size() == 0:
	print('No element found')
else:
	print('\n\nTotal number of elements in linked list is',llist.size())